sdk/nodejs/runtime/closure/createClosure.ts

```diff
@@ -748,23 +751,74 @@ function getOrCreateEntry(
     dispatchAny();
     return entry;
 
-    function dispatchAny(): void {
+    function doNotCapture() {
         if (!serialize(obj)) {
             // caller explicitly does not want us to capture this value.
-            entry.json = undefined;
+            return true;
         }
-        else if (obj && obj.doNotCapture) {
+
+        if (obj && obj.doNotCapture) {
             // object has set itself as something that should not be captured.
-            entry.json = undefined;
+            return true;
         }
-        else if (isDerivedNoCaptureConstructor(obj)) {
+
+        if (isDerivedNoCaptureConstructor(obj)) {
             // this was a constructor that derived from something that should not be captured.
+            return true;
+        }
+
+        return false;
+    }
+
+    function dispatchAny(): void {
+        if (doNotCapture()) {
+            // We do not want to capture this object.  Explicit set .json to undefined so
+            // that we will see that the property is set and we will simply roundtrip this
+            // as the 'undefined value.
             entry.json = undefined;
+            return;
         }
-        else if (obj === undefined || obj === null ||
+
+        if (obj === undefined || obj === null ||
             typeof obj === "boolean" || typeof obj === "number" || typeof obj === "string") {
             // Serialize primitives as-is.
             entry.json = obj;
+            return;
+        }
+
+        const moduleName = findModuleName(obj);
+        if (moduleName && moduleName[0] !== ".") {
```

Comment:
The logic here is a little hard to follow.  I think there are three cases:
1. A local module
2. A non-local module marked with the "no serialize" bit (currently hardcoded to `@pulumi` scope)
3. Everything else

Can the logic here explicitly branch to those three cases, and call out what the precondition is and why each behaves the way it does?

Reply:
Sure.

Reply:
I've extracted a function for handling modules, and i've tried to clearly break out and doucment these cases.